app/domain/authentication/authn_gcp/restrictions_validator.rb

```diff
@@ -0,0 +1,101 @@
+        validate
+        @logger.debug(LogMessages::Authentication::AuthnGCP::ValidatedRestrictionsConstraintCombination.new)
+      end
+
+      private
+
+      def validate
+        validate_constraints_exist
+        validate_restrictions_exist
+        validate_restrictions_types_exist
+        validate_there_are_no_duplicated_constraints
+        validate_there_are_no_duplicated_restrictions
+        validate_constraints_are_permitted
+        validate_restrictions_values_exist
+      end
+
+      def validate_constraints_exist
+        if @permitted_constraints.nil? || @permitted_constraints.empty?
+          raise Errors::Authentication::IllegalStateResourceRestrictionsValidation.new(ILLEGAL_STATE_MISSING_CONSTRAINTS_MESSAGE)
+        end
+      end
+
+      def validate_restrictions_types_exist
+        @resource_restrictions.each do |r|
+          resource_type = r.type
+          if resource_type.nil? || resource_type.empty?
+            raise Errors::Authentication::IllegalStateResourceRestrictionsValidation.new(ILLEGAL_STATE_MISSING_RESOURCE_TYPE_MESSAGE)
+          end
+        end
+      end
+
+      def validate_there_are_no_duplicated_constraints
+        if @permitted_constraints.uniq!
+          raise Errors::Authentication::IllegalStateResourceRestrictionsValidation.new(ILLEGAL_STATE_DUPLICATE_CONSTRAINTS_MESSAGE)
+        end
+      end
+
+      def validate_there_are_no_duplicated_restrictions
+        if !@resource_restrictions.nil? && resource_restrictions_types.uniq!
+          raise Errors::Authentication::IllegalStateResourceRestrictionsValidation.new(ILLEGAL_STATE_DUPLICATE_RESTRICTIONS_MESSAGE)
+        end
+      end
+
+      def validate_restrictions_exist
+        if @resource_restrictions.nil? || @resource_restrictions.empty?
+          raise Errors::Authentication::RoleMissingRequiredConstraints.new(@permitted_constraints)
+        end
+      end
+
+      def validate_constraints_are_permitted
+        @resource_restrictions.each do |r|
+          resource_type = r.type
+          unless @permitted_constraints.include?(resource_type)
+            raise Errors::Authentication::ConstraintNotSupported.new(resource_type, @permitted_constraints)
+          end 
+        end
+      end
+
+      def validate_restrictions_values_exist
+        @resource_restrictions.each do |r|
+          resource_type = r.type
+          resource_value = r.value
+          if resource_value.nil? || resource_value.empty?
+            raise Errors::Authentication::MissingResourceRestrictionsValue.new(resource_type)
+          end
+        end
+      end
+
+      def resource_restrictions_types
+        return @resource_restrictions_types if @resource_restrictions_types
+
+        @resource_restrictions_types = Array.new
+        @resource_restrictions.each do |r|
+          resource_type = r.type
+          next unless resource_type
```

Comment:
will this happen? when will `r` won't have a `type` method?

Reply:
In case of a `r=nil` is already validated before this action,
About getting a different object or a code changing in `resource_restriction` class i think the UT is protecting us.
Do you think we need an additional protection here?

Reply:
i don't think we need to check it at all. you're getting a list of `ResourceRestrictions` objects and all of them have a `type` attribute.

Reply:
i think that you can do this instead:
```
def resource_restrictions_types
  @resource_restrictions_types ||= @resource_restrictions.collect {|r| r.type }
end
```

Reply:
Ohh i am testing here that type is not empty, it is an illegal state error that shouldn't happened.
But i think this class responsibility is also to validate the input parameters

Reply:
why would type be empty? you create the `ResourceRestriction` object and give it value. in which scenario will the type be empty?